### PR TITLE
Use a pip-installed libsupermesh

### PR DESCRIPTION
**Pull request by connorjward**
# Description

I may have figured out the secret sauce.

Merge with https://github.com/firedrakeproject/libsupermesh/pull/11

### TODOs
- [x] Amend pip install information

### CODE REVIEW

File: scripts/firedrake-install [hunk 1]
```diff
@@ -76,7 +76,7 @@ class FiredrakeConfiguration(dict):
                     self["options"][o] = args.__dict__[o]
 
     _persistent_options = ["package_manager",
-                           "minimal_petsc", "mpicc", "mpicxx", "mpif90", "mpiexec", "disable_ssh",
+                           "minimal_petsc", "mpicc", "mpicxx", "mpif90", "mpiexec", "mpihome", "disable_ssh",
```
Comment (JDBetteridge) [inline scripts/firedrake-install:79]: Can the mpi* options here be replaced with mpi_dir or similar (which would mirror the PETSc configure argument)?

Comment (connorjward) [reply scripts/firedrake-install:79]: I agree that that would be nicer. But I think this is simpler to implement and `firedrake-install` might well die once this PR is merged.

**Comment by francesco-ballarin** [comment]
Almost unrelated question, apologies for having missed the right time to ask: why are the `requirements*.txt` files still around? I thought everything had moved into `pyproject.toml`.

**Comment by connorjward** [comment]
> Almost unrelated question, apologies for having missed the right time to ask: why are the `requirements*.txt` files still around? I thought everything had moved into `pyproject.toml`.

They are parsed inside the `firedrake-install` script. Since `firedrake-install` will hopefully be going away very soon I don't want to spend a lot of time updating it which is why I have left them.

File: scripts/firedrake-install (hunk 1 above)
Comment (JDBetteridge) [reply scripts/firedrake-install:79]: Maybe we can sniff this new value from `args.mpiexec`

Comment (connorjward) [reply scripts/firedrake-install:79]: This will not work as sometimes we pass things like `--mpicc mpicc` (i.e. not a path).

File: setup.py [hunk 2]
```diff
@@ -159,15 +163,16 @@ def __getitem__(self, key):
...
+# gcc path/to/libsupermesh/libsupermesh.cpython-311-x86_64-linux-gnu.so \
 #    -lsupermesh \
-#    -Wl,-rpath,$ORIGIN/../../supermesh
-supermesh_ = ExternalDependency(
-    include_dirs=[f"{sys.prefix}/include"],
-    library_dirs=[f"{sys.prefix}/lib"],
+#    -Wl,-rpath,$ORIGIN/../../libsupermesh
+libsupermesh_dir = Path(libsupermesh.__path__._path[0]).absolute()
+libsupermesh_ = ExternalDependency(
+    include_dirs=[str(libsupermesh_dir.joinpath("include"))],
```
Comment (connorjward) [inline setup.py:172]:
```suggestion
    include_dirs=[libsupermesh.get_include()],
```

Comment (connorjward) [inline setup.py:172]:
```suggestion
    include_dirs=[libsupermesh.get_include()],
```

File: setup.py [hunk 3]
```diff
@@ -159,15 +163,16 @@ def __getitem__(self, key):
...
+    library_dirs=[str(libsupermesh_dir.joinpath("lib"))],
```
Comment (connorjward) [inline setup.py:173]:
```suggestion
    library_dirs=[str(Path(libsupermesh.get_library()).parent)],
```

File: setup.py [hunk 4]
```diff
@@ -159,15 +163,16 @@ def __getitem__(self, key):
...
-# gcc /supermesh/supermesh.cpython-311-x86_64-linux-gnu.so \
+# gcc -Ipath/to/libsupermesh/include
+# gcc path/to/libsupermesh/libsupermesh.cpython-311-x86_64-linux-gnu.so \
 #    -lsupermesh \
-#    -Wl,-rpath,$ORIGIN/../../supermesh
-supermesh_ = ExternalDependency(
-    include_dirs=[f"{sys.prefix}/include"],
-    library_dirs=[f"{sys.prefix}/lib"],
+#    -Wl,-rpath,$ORIGIN/../../libsupermesh
+libsupermesh_dir = Path(libsupermesh.__path__._path[0]).absolute()
```
Comment (connorjward) [inline setup.py:170]:
```suggestion
```

File: firedrake/supermeshing.py [hunk 5]
```diff
@@ -428,7 +429,8 @@ def likely(cell_A):
         "complex_mode": 1 if complex_mode else 0
     }
 
-    dirs = get_petsc_dir() + (sys.prefix, )
+    libsupermesh_dir = pathlib.Path(libsupermesh.__path__._path[0]).absolute()
```
Comment (connorjward) [inline firedrake/supermeshing.py:432]:
```suggestion
```

File: pyproject.toml [hunk 6]
```diff
@@ -38,6 +38,7 @@ dependencies = [
   "fenics-fiat @ git+https://github.com/firedrakeproject/fiat.git",
   "pyadjoint-ad @ git+https://github.com/dolfin-adjoint/pyadjoint.git",
   "loopy @ git+https://github.com/firedrakeproject/loopy.git@main",
+  "libsupermesh @ git+https://github.com/firedrakeproject/libsupermesh.git@connorjward/pip-install2",
```
Comment (connorjward) [inline pyproject.toml:41]:
```suggestion
  "libsupermesh @ git+https://github.com/firedrakeproject/libsupermesh.git",
```

File: pyproject.toml [hunk 7]
```diff
@@ -98,6 +99,7 @@ requires = [
   "mpi4py; python_version < '3.13'",
   "petsc4py",
   "rtree>=1.2",
+  "libsupermesh @ git+https://github.com/firedrakeproject/libsupermesh.git@connorjward/pip-install2",
```
Comment (connorjward) [inline pyproject.toml:102]:
```suggestion
  "libsupermesh @ git+https://github.com/firedrakeproject/libsupermesh.git",
```

File: requirements-git.txt [hunk 8]
```diff
@@ -4,3 +4,4 @@ git+https://github.com/dolfin-adjoint/pyadjoint.git#egg=pyadjoint-ad
 git+https://github.com/firedrakeproject/loopy.git@main#egg=loopy
 git+https://github.com/firedrakeproject/pytest-mpi.git@main#egg=pytest-mpi
 git+https://github.com/firedrakeproject/petsc.git@firedrake#egg=petsc
+git+https://github.com/firedrakeproject/libsupermesh.git@connorjward/pip-install2
```
Comment (connorjward) [inline requirements-git.txt:7]:
```suggestion
git+https://github.com/firedrakeproject/libsupermesh.git#egg=libsupermesh
```
unsure about this egg business

Comment (JDBetteridge) [reply requirements-git.txt:7]: It just corresponds to the name of the underlying package, which may be different to the repo name. This is fine!

**Review by JDBetteridge — APPROVED** [review]
Looks good, let's get this in!

**Pull request merged**